Add `undo` and `redo` command functionality

CJerrong opened the pull request:
Main changes in this pr:
- Support new undo and redo commands.
- Added a new datastoreVersionStorage which is in charge of storing snapshots of the datastore whenever it is mutated.
- Refactor commands that mutate the datastore to commit the datastore to the datastoreVersionStorage after completion.
- Update the developer guide regarding implementation of undo and redo functionality

CJerrong committed:
Add DatastoreVersionStorage class to store datastore snapshots

CJerrong committed:
Add undoCommand and redoCommand

CJerrong committed:
Refactor code to abide with checkstyle requirements

chaaaaun commented on src/main/java/scrolls/elder/model/DatastoreVersionStorage.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    public ReadOnlyDatastore executeUndo() {
+        assert this.canUndo() : "Undo command cannot be executed";
```
Actually this prompted me to do a little searching. According to [this post](https://stackoverflow.com/questions/2758224/what-are-assertions-in-java-and-when-should-they-be-used), assertions aren't even turned on by default. Should we operate on the assumption that they don't trigger?

gerteck replied:
Agreed. If I understand this correctly, the logic here will still hold, since `UndoCommand::Execute` has a guard clause checking `canUndo`, which throws an exception if it fails.  Hence, the assertion here is a supplementary sanity check to ensure correct usage of the `executeUndo` method 👍

chaaaaun commented on src/main/java/scrolls/elder/model/DatastoreVersionStorage.java:
```diff
@@ -50,7 +50,16 @@ public ReadOnlyDatastore executeRedo() {
-        this.datastoreVersions.add(datastore);
+        int size = this.datastoreVersions.size();
+
+        // If not at end of list, purge the data before adding new datastore snapshot
```
Good catch, we might need a couple of tests for various sequences of undo and redo to catch any edge cases.

CJerrong replied:
Yep I will be implementing test cases on undo and redo commands and pushing them to this pr.

chaaaaun commented on src/main/java/scrolls/elder/logic/commands/PairCommand.java:
```diff
@@ -118,6 +118,7 @@ public CommandResult execute(Model model) throws CommandException {
         store.setPerson(personToPair1, newPerson1);
         store.setPerson(personToPair2, newPerson2);
+        model.commitDatastore();
         store.updateFilteredPersonList(Model.PREDICATE_SHOW_ALL);
```
Would it be better to commit the Datastore after the personlist is reset? Then when someone undos they get the same view that they see when they previously ran the command.

chaaaaun reviewed (approved):
Overall good clean implementation!

soons1 commented on src/main/java/scrolls/elder/logic/commands/RedoCommand.java:
```diff
@@ -0,0 +1,25 @@
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        requireNonNull(model);
+        if (!model.getDatastoreVersionStorage().canRedo()) {
```
Checking if there is a previous operation to be redone is important, but how might that tie in with the assert statement in redoChanges() in terms of 2 layers of checking?

CJerrong replied:
I treated the assert statement in redoChanges more like a sanity check to ensure correctness, although it might be fine to remove the assertion as well

soons1 commented on src/main/java/scrolls/elder/logic/commands/RedoCommand.java:
```diff
@@ -0,0 +1,25 @@
+            throw new CommandException(MESSAGE_REDO_ERROR);
+        }
+        model.redoChanges();
+        return new CommandResult(MESSAGE_SUCCESS);
```
Should we consider updating the displayed lists after redoing the changes?

soons1 commented on src/main/java/scrolls/elder/logic/commands/UndoCommand.java:
```diff
@@ -0,0 +1,26 @@
+        }
+
+        model.undoChanges();
+        return new CommandResult(MESSAGE_SUCCESS);
```
Same as RedoCommand, should we update the displayed list after undoing the operation?

CJerrong replied:
That's a good point, maybe it would be be clearer for the user as well, I'll add that in.

soons1 commented on src/main/java/scrolls/elder/model/DatastoreVersionStorage.java:
```diff
@@ -0,0 +1,109 @@
+     * @return The Datastore in its next state after reversing an undo operation.
+     */
+    public ReadOnlyDatastore executeRedo() {
+        assert this.canRedo() : "Redo command cannot be executed";
```
If we were to fall to this layer of assertion, is it possible to display the message string to the user on the GUI?

soons1 commented on src/main/java/scrolls/elder/logic/parser/AddressBookParser.java:
```diff
@@ -92,6 +94,12 @@ public Command parseCommand(String userInput) throws ParseException {
         case LogAddCommand.COMMAND_WORD:
             return new LogAddCommandParser().parse(arguments);
 
+        case UndoCommand.COMMAND_WORD:
```
Do we want to consider alternative command words for the UndoCommand? We can possibly consider "reverse", "cancel" etc.

CJerrong replied:
That's a possible alternative as well, maybe we can discuss it and I can make the changes accordingly.

soons1 reviewed (commented):
LGTM 👍 
Just need to add some test cases to validate some functions in ModelManager, RedoCommand, UndoCommand etc. and it is good to go!

gerteck reviewed (approved):
LGTM, would be good to add supplementary testcases and it will be ready to merge! 🍔

CJerrong committed:
Refactor commands to refresh the filtered lists after completion

CJerrong committed:
Add tests for undoCommand and redoCommand

CJerrong committed:
Refactor test cases to abide by checkstyle requirements

CJerrong committed:
Refactor UndoSequenceDiagrams to adhere to current design

CJerrong committed:
Update the DeveloperGuide on the new undo and redo command functionality

Update the diagrams and images involved in illustrating the undo and
redo mechanisms

CJerrong committed:
Merge branch 'master' into branch-add-undo

gerteck commented on src/test/java/scrolls/elder/logic/commands/UndoCommandTest.java:
```diff
@@ -0,0 +1,158 @@
+    private PersonStore personStore;
+    private PersonStore expectedPersonStore;
+
+    @BeforeEach
```
Nice Usage of `@BeforeEach` here

gerteck reviewed (approved):
Tests LGTM 👍
